### PR TITLE
Performance

### PR DESCRIPTION
**Pull request by janisozaur**
Includes #2979 

I converted all of the control-flow dependencies in inner-most loop in `DrawRLESprite2` into data-dependencies. The main purpose of this is to remove branching, which can prevent OOO execution on CPU.

Don't be fooled by the few remaining `if`s are actually compile-time constants due to #2979's introduction of templating.

This has minor effect on performance for me, but should come in handy in case someone does a GPU renderer.

Requires testing.

### CODE REVIEW

File: src/drawing/drawing_fast.cpp [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+            //Shorten the line
+            no_pixels -= pixels_till_end * greater_than_zero(pixels_till_end);
+
+            //Finally after all those checks, copy the image onto the drawing surface
+            //If the image type is not a basic one we require to mix the pixels
+            if (image_type & IMAGE_TYPE_USE_PALETTE) {//In the .exe these are all unraveled loops
+                for (; no_pixels > 0; no_pixels -= zoom_amount, source_pointer += zoom_amount, dest_pointer++) {
+                    uint8 al = *source_pointer;
+                    uint8 ah = *dest_pointer;
+                    if (image_type & IMAGE_TYPE_MIX_BACKGROUND)
```
Comment (duncanspumpkin) [inline src/drawing/drawing_fast.cpp]: We should branch this earlier before the `for` if we are making a fast version.

**Comment by IntelOrca** [comment]
I looked at the assembly, it has reduced the number of (conditional) jump instructions considerably in return for more arithmetic instructions. This may improve performance as there is less need to flush pipeline.

I however have not noticed an FPS increase. I think this function is not going to get any faster. We should look elsewhere to optimise.

Before: http://hastebin.com/kudecotoxa.asm
After: http://hastebin.com/uqowejayeg.asm

File: src/drawing/drawing_fast.cpp (hunk 1 above)
Comment (IntelOrca) [inline src/drawing/drawing_fast.cpp]: The if statement is not compiled.

File: src/drawing/drawing_fast.cpp [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    int line_width = (dpi->width >> zoom_level) + dpi->pitch;
+
+    if (source_y_start < 0) {
+        source_y_start += zoom_amount;
+        next_dest_pointer += line_width;
+        height -= zoom_amount;
+    }
+
+    //For every line in the image
+    for (int y = source_y_start; y < (height + source_y_start); y += zoom_amount) {
```
Comment (IntelOrca) [inline src/drawing/drawing_fast.cpp]: We could store ```height + source_y_start```, but I am not sure if it will improve performance or not.

Comment (janisozaur) [inline src/drawing/drawing_fast.cpp]: I'm quite certain compiler optimises it anyway.

**Comment by janisozaur** [comment]
I've just had a chance to test it out, it does bring around (at most) +1 fps (against #2979) when compiled in release mode. Since it doesn't decrease performance and possibly makes GPU version a tad easier, I vote for this to be merged – if #2979 is to be merged.

File: src/drawing/drawing_fast.cpp (hunk 2 above)
Comment (IntelOrca) [inline src/drawing/drawing_fast.cpp]: How are you certain, just a hunch?

Comment (janisozaur) [inline src/drawing/drawing_fast.cpp]: yes, but I can make the change you suggest nevertheless.

File: src/drawing/drawing_fast.cpp [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        }
+        else if (image_type & IMAGE_TYPE_MIX_BACKGROUND)
+        {
+            DrawRLESpriteHelper1(IMAGE_TYPE_MIX_BACKGROUND);
+        }
+        else
+        {
+            DrawRLESpriteHelper1(0);
+        }
+    }
```
Comment (janisozaur) [inline src/drawing/drawing_fast.cpp]: wouldn't it suffice to call `helper1(image_type & (MIX_BACKGROUND | PALLETTED))` here? or does that confuse the compiler?

Comment (IntelOrca) [inline src/drawing/drawing_fast.cpp]: Template parameters must be constants or constant expressions.

**Comment by janisozaur** [comment]
One reason for missing speedup here is compiler generating `imul` (latency 3 or 4 for haswell) rather than some broadcast. Perhaps it believes `sign` can be anything rather than just 1 or 0, I'll look into that.

I made the change you mentioned @IntelOrca .

**Comment by janisozaur** [comment]
Ok, I've updated the code, can you check now?

I'm sure there is more performance to be squeezed out of this bit, though getting there will only be trickier and trickier.

File: src/drawing/drawing_fast.cpp [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        //The first part of the source pointer is a list of offsets to different lines
+        //This will move the pointer to the correct source line.
+        const uint8 *next_source_pointer = source_bits_pointer + ((uint16*)source_bits_pointer)[y];
+
+        uint8 last_data_line = 0;
+
+        //For every data section in the line
+        while (!last_data_line) {
+            const uint8* source_pointer = next_source_pointer;
+            uint8* dest_pointer = next_dest_pointer + line_width * i2;
```
Comment (IntelOrca) [inline src/drawing/drawing_fast.cpp]: You have added a multiplication to inside a loop... is this really an optimisation?

File: src/drawing/drawing_fast.cpp [hunk 5]
```diff
@@ -0,0 +1,186 @@
+extern "C"
+{
+    #include "drawing.h"
+}
+
+// This will have -1 (0xffffffff) for (val <= 0), 0 otherwise, so it can act as a mask
+// This is expected to generate
+//     sar $0x1f, %eax (arithmetic shift right by 31)
```
Comment (duncanspumpkin) [inline src/drawing/drawing_fast.cpp]: stick to intel syntax for comments.

Comment (IntelOrca) [inline src/drawing/drawing_fast.cpp]: Wouldn't it be ```sar eax, 01Fh```?

File: src/drawing/drawing_fast.cpp (hunk 4 above)
Comment (janisozaur) [inline src/drawing/drawing_fast.cpp]: I moved it one level higher, it was not meant to end up here

File: src/drawing/drawing_fast.cpp (hunk 5 above)
Comment (janisozaur) [inline src/drawing/drawing_fast.cpp]: fixed

**Comment by janisozaur** [comment]
I updated the last commit.

Still requires some testing + review, though I checked it quite extensively on my end.

**Pull request merged**